CMakeLists.txt

```diff
@@ -251,15 +251,18 @@ if(DEACTIVATE_AVX2)
 endif()
 
 # flags
-# @TODO: set -Wall
+# Set -Wall and other useful warning flags.
+if(CMAKE_C_COMPILER_ID STREQUAL GNU OR CMAKE_C_COMPILER_ID STREQUAL Clang OR CMAKE_C_COMPILER_ID STREQUAL Intel)
```

Comment:
You could also toss `AppleClang` in there for good measure. It's how Apple's own version of Clang is identified.

Instead of the `if`, you could use generator expressions like:

```cmake
add_compile_options(
    "$<$<CXX_COMPILER_ID:GNU>:-Wall;-Wwrite-strings;-Wno-unused>"
    "$<$<CXX_COMPILER_ID:Clang>:-Wall;-Wwrite-strings;-Wno-unused>"
    "$<$<CXX_COMPILER_ID:AppleClang>:-Wall;-Wwrite-strings;-Wno-unused>"
    "$<$<CXX_COMPILER_ID:Intel>:-Wall;-Wwrite-strings;-Wno-unused>"
)
```

But I don't know if that's any prettier? Best-looking would be if the 3.3+ feature `if ("bar" IN_LIST _list)` could be used I think, but that's not in 2.8.

If you keep it the way it is, perhaps move the `add_compile_option` to inside the `if` block just below (the one for SSE2)? (since the conditions are the same).

Reply:
In fact, thinking a bit, don't take the generator expressions as a serious suggestion. It's just cryptic, while anyone who reads the `CMakeLists.txt` probably just want to get on with what they're doing and not learn all about CMake's warts :)

Generator expressions seems to be primarily for when you need information that is only available at CMake's final "generation" step, and the compiler ID is not such a thing.